src/redis-cli.c

```diff
@@ -1930,22 +1934,43 @@ static int confirmWithYes(char *msg, int ignore_force) {
     return (nread != 0 && !strcmp("yes", buf));
 }
 
-/* Turn the plain C strings into Sds strings */
-static char **convertToSds(int count, char** args) {
-    int j;
-    char **sds = zmalloc(sizeof(char*)*count);
+/* Unquote a null-terminated string and return it as a binary-safe sds. */
+static sds unquoteCString(char *str) {
+    int count;
+    sds *unquoted = sdssplitargs(str, &count);
 
-    for(j = 0; j < count; j++)
-        sds[j] = sdsnew(args[j]);
+    if (unquoted && count == 1) {
+        sds res = unquoted[0];
+        sds_free(unquoted);
+        return res;
+    }
 
-    return sds;
+    if (unquoted)
+        sdsfreesplitres(unquoted, count);
+    return NULL;
 }
 
-static void freeConvertedSds(int count, char **sds) {
-    int j;
-    for (j = 0; j < count; j++)
-        sdsfree(sds[j]);
-    zfree(sds);
+/* Create an sds array from argv, either as-is or by dequoting every
+ * element. When quoted is non-zero, may return a NULL to indicate an
+ * invalid quoted string.
+ */
+static sds *getSdsArrayFromArgv(int argc, char **argv, int quoted) {
+    sds *res = sds_malloc(sizeof(sds) * argc);
+
+    for (int j = 0; j < argc; j++) {
+        if (quoted) {
+            sds unquoted = unquoteCString(argv[j]);
+            if (!unquoted) {
+                while (--j > 0) sdsfree(res[j]);
+                return NULL;
```

Comment:
```suggestion
                s_free(res);
                return NULL;
```
Shall we also free the allocated array here?